bgv/encoder.go

```diff
@@ -92,254 +60,422 @@ func NewEncoder(params Parameters) Encoder {
+		bufT:        ringT.NewPoly(),
+		bufB:        bufB,
 		paramsQP:    paramsQP,
 		qHalf:       qHalf,
 		tInvModQ:    tInvModQ,
 	}
 }
 
-// EncodeNew encodes a slice of integers of type []uint64 or []int64 of size at most N on a newly allocated plaintext.
-func (ecd *encoder) EncodeNew(values interface{}, level int, scale rlwe.Scale) (pt *rlwe.Plaintext) {
-	pt = NewPlaintext(ecd.params, level)
-	pt.Scale = scale
-	ecd.Encode(values, pt)
-	return
-}
+func permuteMatrix(logN int) (perm []uint64) {
 
-// Encode encodes a slice of integers of type []uint64 or []int64 of size at most N into a pre-allocated plaintext.
-func (ecd *encoder) Encode(values interface{}, pt *rlwe.Plaintext) {
-	ecd.EncodeRingT(values, pt.Scale, ecd.buffT)
-	ecd.RingT2Q(pt.Level(), ecd.buffT, pt.Value)
+	var N, pow, pos uint64 = uint64(1 << logN), 1, 0
 
-	if pt.IsNTT {
-		ecd.params.RingQ().AtLevel(pt.Level()).NTT(pt.Value, pt.Value)
-	}
+	mask := 2*N - 1
 
-	ecd.ScaleUp(pt.Level(), pt.Value, pt.Value)
-}
+	perm = make([]uint64, N)
 
-// EncodeCoeffs encodes a slice of []uint64 of size at most N on a pre-allocated plaintext.
-// The encoding is done coefficient wise, i.e. [1, 2, 3, 4] -> 1 + 2X + 3X^2 + 4X^3.
-func (ecd *encoder) EncodeCoeffs(values []uint64, pt *rlwe.Plaintext) {
-	copy(ecd.buffT.Coeffs[0], values)
+	halfN := int(N >> 1)
 
-	for i := len(values); i < len(ecd.buffT.Coeffs[0]); i++ {
-		ecd.buffT.Coeffs[0][i] = 0
-	}
+	for i, j := 0, halfN; i < halfN; i, j = i+1, j+1 {
 
-	ringT := ecd.params.RingT()
+		pos = utils.BitReverse64(pow>>1, logN) // = (pow-1)/2
 
-	ringT.MulScalar(ecd.buffT, pt.Scale.Uint64(), ecd.buffT)
-	ecd.RingT2Q(pt.Level(), ecd.buffT, pt.Value)
+		perm[i] = pos
+		perm[j] = N - pos - 1
 
-	if pt.IsNTT {
-		ecd.params.RingQ().AtLevel(pt.Level()).NTT(pt.Value, pt.Value)
+		pow *= GaloisGen
+		pow &= mask
 	}
 
-	ecd.ScaleUp(pt.Level(), pt.Value, pt.Value)
+	return perm
 }
 
-// EncodeCoeffsNew encodes a slice of []uint64 of size at most N on a newly allocated plaintext.
-// The encoding is done coefficient wise, i.e. [1, 2, 3, 4] -> 1 + 2X + 3X^2 + 4X^3.}
-func (ecd *encoder) EncodeCoeffsNew(values []uint64, level int, scale rlwe.Scale) (pt *rlwe.Plaintext) {
-	pt = NewPlaintext(ecd.params, level)
-	pt.Scale = scale
-	ecd.EncodeCoeffs(values, pt)
-	return
+// GetRLWEParameters returns the underlying rlwe.Parameters of the target object.
+func (ecd Encoder) GetRLWEParameters() *rlwe.Parameters {
+	return &ecd.parameters.Parameters
 }
 
-// EncodeRingT encodes a slice of []uint64 or []int64 on a polynomial in basis T.
-func (ecd *encoder) EncodeRingT(values interface{}, scale rlwe.Scale, pT *ring.Poly) {
+// Encode encodes an IntegerSlice of size at most N, where N is the smallest value satisfying PlaintextModulus = 1 mod 2N,
+// on a pre-allocated plaintext.
+func (ecd Encoder) Encode(values IntegerSlice, pt *rlwe.Plaintext) (err error) {
+
+	if pt.IsBatched {
+		return ecd.Embed(values, true, pt.MetaData, pt.Value)
+	} else {
```

Comment:
The part of the Encode and Decode methods when IsBatched == false aren't tested in `bgv_test.go`
But I'm not sure how to create such test.

Reply:
I've taken care of it